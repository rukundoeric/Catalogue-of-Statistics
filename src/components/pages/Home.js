import React, { useState, useEffect } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import Banner from '../layouts/Banner';
import Meals from '../layouts/Meals';
import MealInfo from '../layouts/MealInfo';
import { getCategories, getRandomMeal, getMeals } from '../../redux/actions';

const Home = ({
  getCategories,
  getRandomMeal,
  getMeals,
  categories: catList,
  randomMeal: rMeal,
  mealsList: mList,
}) => {
  const [categories, setCategories] = useState([]);
  const [randomMeal, setRandomMeal] = useState(undefined);
  const [mealsList, setMealsList] = useState([]);

  useEffect(() => { getCategories(); getRandomMeal(); getMeals(); }, []);
  useEffect(() => { setCategories(catList); }, [catList]);
  useEffect(() => { setRandomMeal(rMeal); }, [rMeal]);
  useEffect(() => { setMealsList(mList); }, [mList]);

  return (
    <div>
      <Banner randomMeal={randomMeal} />
      <MealInfo meal={randomMeal} />
      <Meals categories={categories} mealsList={mealsList} />
    </div>
  );
};

Home.propTypes = {
  getRandomMeal: PropTypes.func.isRequired,
  getCategories: PropTypes.func.isRequired,
  getMeals: PropTypes.func.isRequired,
  categories: PropTypes.arrayOf(PropTypes.shape({})).isRequired,
  randomMeal: PropTypes.shape({}).isRequired,
  mealsList: PropTypes.arrayOf(PropTypes.shape({})).isRequired,
};

const mapStateToProps = ({ meals: { mealsList, randomMeal }, categories }) => ({
  mealsList,
  randomMeal,
  categories,
});

const mapDispatchToProps = {
  getRandomMeal,
  getCategories,
  getMeals,
};

export default connect(mapStateToProps, mapDispatchToProps)(Home);
