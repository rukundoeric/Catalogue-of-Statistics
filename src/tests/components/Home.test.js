import '@testing-library/jest-dom';
import { BrowserRouter } from 'react-router-dom';
import { screen, waitFor, fireEvent } from '@testing-library/react';
import server from '../helpers/Server';
import Home from '../../components/pages/Home';
import Render from '../helpers/Render';

const state = {
  initialState: {
    meals: {
      randomMeal: {
        strMeal: 'ab',
        strInstructions: 'bc',
        strIngredient1: 'a',
        strIngredient2: 'a',
        strIngredient3: '',
      },
      mealDetails: {},
      mealsList: [{
        strMeal: 'ab',
        strInstructions: 'bc',
        strIngredient1: 'a',
        strIngredient2: 'a',
        strIngredient3: '',
      }, {
        strMeal: 'ab',
        strInstructions: 'bc',
        strIngredient1: 'a',
        strIngredient2: 'a',
        strIngredient3: '',
      }],
    },
    categories: [
      {
        strCategory: 'Select category',
      },
      {
        strCategory: 'ab',
      },
      {
        strCategory: 'ac',
      },
    ],
  },
};

describe('Home :', () => {
  beforeAll(() => server.listen());
  afterEach(() => server.resetHandlers());
  afterAll(() => server.close());

  it('Should render Home page', async () => {
    Render(
      <BrowserRouter>
        <Home />
      </BrowserRouter>, { ...state },
    );

    await waitFor(() => screen.getByTestId('meal-name'));

    expect(screen.getByTestId('meal-name')).toHaveTextContent('ab');
    expect(screen.getByTestId('meal-instruction')).toHaveTextContent('bc');
  });

  it('Should search meal', async () => {
    Render(
      <BrowserRouter>
        <Home />
      </BrowserRouter>, { ...state },
    );

    await waitFor(() => screen.getByTestId('search-input'));

    fireEvent.change(screen.getByTestId('search-input'), { target: { value: 'a' } });
    expect(screen.getByTestId('search-input')).toHaveValue('a');
    fireEvent.change(screen.getByTestId('search-input'), { target: { value: '' } });
    expect(screen.getByTestId('search-input')).toHaveValue('');
  });

  it('Should filter by category', async () => {
    Render(
      <BrowserRouter>
        <Home />
      </BrowserRouter>, { ...state },
    );

    await waitFor(() => screen.getByTestId('filter-select'));

    fireEvent.change(screen.getByTestId('filter-select'), { target: { value: 'ab' } });
    expect(screen.getByTestId('filter-select')).toHaveValue('ab');
    fireEvent.change(screen.getByTestId('filter-select'), { target: { value: 'Select category' } });
    expect(screen.getByTestId('filter-select')).toHaveValue('Select category');
  });
});
