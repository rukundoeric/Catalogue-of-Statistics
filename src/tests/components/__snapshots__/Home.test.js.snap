// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should match snapshot 1`] = `
<DocumentFragment>
  <div>
    <h1>
      Home
    </h1>
  </div>
</DocumentFragment>
`;
